GSA-TTS/FAC
# Super Simple Search

# Simple Search - #1170

## Changes
`/search` is a new route, under a new app, with some functionality.

1. `/search` is the search home page, linked to in the header. It contains the simple search functionality with a few fields. Form works, but currently does nothing but redirect to itself. 
   - UEI or EIN
   - ALN/CFDA
   - Fiscal Year
   - Acceptance date

2. `/search/table?query` gets and displays search results from the API. This string works **one-to-one** with the API, with two exceptions. 
   - This means a request made to
   `http://localhost:8000/search/vw_auditee?limit=20&auditee_city=eq.ARLINGTON` 
   should display the exact same results as we would get in JSON by requesting from 
   `https://api.data.gov/TEST/audit-clearinghouse/v0/dissemination/vw_auditee?limit=20&auditee_city=eq.ARLINGTON`
   - The exceptions: visiting pages made with `/vw_general` or `/vw_federal_award` will trigger an extra query. This extra query cross-references the previous results with the `/vw_auditee` table, so as to gain access to things like `auditee_city`
      - This means that requests to other endpoints will not display correctly on the front end, which is fine. It is receiving the data, just not the kind it wants to show.
      - This will be changed in the future on a layer further back than Django, where a simple "LEFT JOIN" can solve the problem. It will also make sorting easier, yay!

3. There exists a session-level variable that can tell whether you've accepted the Uniform Guidance Acknowledgement. The modal that controls this variable is available on the bottom of the search results. In the future, it will pop up in front of users once, and because of this variable, will not be seen again. 

## How to test
1. Follow the instructions to get an api key [here](https://github.com/GSA-TTS/FAC/discussions/1066). Once approved, define it as `API_GOV_KEY` wherever you keep your environment variables. Also, set `API_GOV_URL` to `"https://api.data.gov/TEST/audit-clearinghouse/v0/dissemination"`. Eventually, this will not be needed, as Django will ping the local database directly. For now, we use the online API. 
2. Swap to this branch and `docker compose up`. It shouldn't require a rebuild or anything (unless you haven't in a while, in which case the postgreSQL update will be in the way).
3. Check out [http://localhost:8000/search/](http://localhost:8000/search/). Here are some searches you can try:
    - Sort by "UEI", and enter `HWG4ML2D9EL6`. One result. [Link](http://localhost:8000/search/vw_auditee?uei_list=cd.%7B%20HWG4ML2D9EL6%20%7D&limit=100)
    - Sort by "EIN", and enter `730785251`. One result. [Link](http://localhost:8000/search/vw_auditee?ein_list=cd.%7B%20730785251%20%7D&limit=100)
    - Sort by "Audit Year", and enter `2022`. One hundred different results. [Link](http://localhost:8000/search/vw_general?audit_year=eq.2022&limit=100)
    - Sort by "ALN", and enter `93.600`. One hundred results, with many repeats. [Link](http://localhost:8000/search/vw_federal_award?agency_cfda=eq.93.600&limit=100)
4. Also try some other things, like manually changing the "limit" value in the query. It shouldn't let you go above 1000, because that did give some response time errors. 

## Things to be done for this PR (probably/maybe):
1. GET AN API KEY FOR STAGING/DEV TO USE BEFORE MERGING
2. Fix formatting on things, finalize language.
3. Include the search component on the results page.
4. Show currently used filters in the results.
5. Make the modal pop up on pageload, if the user's not seen it before.

Edit: Updated search filters. 

## Screenshots
![image](https://github.com/GSA-TTS/FAC/assets/91098850/7e872297-c95a-44bd-af80-298dcc906884)
![image](https://github.com/GSA-TTS/FAC/assets/91098850/f62adb11-275e-4a5d-953a-138769458837)
![image](https://github.com/GSA-TTS/FAC/assets/91098850/dbecaaa2-f85e-4792-aaa0-75c6ed3647f7)
